src/scripts/orb.py

```diff
@@ -0,0 +1,145 @@
+bash_env_path = os.environ.get('BASH_ENV')
+if not bash_env_path:
+    exit('ERROR: missing BASH_ENV environment variable')
+
+# Constants
+org_name = os.environ.get("CIRCLE_PROJECT_USERNAME")
+repo = os.environ.get("CIRCLE_PROJECT_REPONAME")
+branch = os.environ.get("CIRCLE_BRANCH")
+
+# Get auth token
+api_token = os.environ.get('SHIPYARD_API_TOKEN')
+if not api_token:
+    exit('No SHIPYARD_API_TOKEN provided, exiting.')
+
+# Prepare API client
+configuration = swagger_client.Configuration()
+configuration.api_key['x-api-token'] = api_token
+client = swagger_client.ApiClient(configuration)
+api_instance = swagger_client.EnvironmentApi(client)
+
+
+def fetch_shipyard_environment():
+    """Fetch the Shipyard environment for this CircleCI job"""
+
+    # Hit the Shipyard API
+    try:
+        response = api_instance.list_environments(org_name=org_name,
+                                                  repo_name=repo,
+                                                  branch=branch).to_dict()
+    except ApiException as e:
+        exit("ERROR: issue while listing environments via API: {}".format(e))
+
+    # Exit if any errors
+    errors = response.get('errors')
+    if errors:
+        exit('ERROR: {}'.format(errors[0]["title"]))
+
+    # Verify an environment was found
+    if not len(response['data']):
+        exit('ERROR: no matching Shipyard environment found')
+
+    # Verify the data is where we expect
+    try:
+        environment_id = response['data'][0]['id']
+        environment_data = response['data'][0]['attributes']
+    except Exception:
+        exit('ERROR: invalid response data structure')
+
+    # Verify all the needed fields are available
+    for param in ('bypass_token', 'url', 'ready', 'stopped', 'retired'):
+        if param not in environment_data:
+            exit('ERROR: no {} found!'.format(param))
+
+    return environment_id, environment_data
+
+
+def restart_environment(environment_id):
+    """Restart the Shipyard environment with the provided ID"""
+
+    try:
+        api_instance.restart_environment(environment_id)
+    except ApiException as e:
+        exit("ERROR: issue while restart the environment: {}".format(e))
+
+
+def wait_for_environment():
+    """Return the Shipyard environment data once it's ready"""
+
+    auto_restart = True
+    was_restarted = False
+
+    # Check the environment
+    environment_id, environment_data = fetch_shipyard_environment()
+
+    # Until the environment is ready
+    while not environment_data['ready']:
+        # Auto-restart the environment once if indicated
+        if all([environment_data['retired'], auto_restart, not was_restarted]):
+            restart_environment(environment_id)
+            was_restarted = True
+            print('Restarted Shipyard environment...')
+        elif environment_data['stopped'] and not environment_data['processing']:
+            exit('ERROR: this environment is stopped and no builds are processing')
```

Comment:
```suggestion
        if all([environment_data['retired'], auto_restart, not was_restarted]):
            if auto_restart and not was_restarted:
                restart_environment(environment_id)
                was_restarted = True
                print('Restarted Shipyard environment...')
            else:
                exit('ERROR: environment is retired and not responding')
        elif environment_data['stopped'] and not environment_data['processing']:
            exit('ERROR: this environment is stopped and no builds are processing')
```
The way I understand this, it will try to restart once. If the environment is not responsive after that, should we break from the loop?

Reply:
It will try to restart once, and if the environment is still retired and nothing is happening, then it will quit on the next loop.

Reply:
Ah right, retired will count as STOPPED.